ankimorphs/highlight_morphs_jit.py

```diff
@@ -0,0 +1,149 @@
+from __future__ import annotations
+
+import re
+
+import anki
+from anki.template import TemplateRenderContext
+
+from . import (
+    ankimorphs_config,
+    ankimorphs_globals,
+    text_highlighting,
+    text_preprocessing,
+)
+from .ankimorphs_config import AnkiMorphsConfig, AnkiMorphsConfigFilter
+from .ankimorphs_db import AnkiMorphsDB
+from .morpheme import Morpheme
+from .morphemizers import morphemizer as morphemizer_module
+from .morphemizers import spacy_wrapper
+from .morphemizers.morphemizer import Morphemizer, SpacyMorphemizer
+
+
+def highlight_morphs_jit(
+    field_text: str,
+    field_name: str,
+    filter_name: str,
+    context: TemplateRenderContext,
+) -> str:
+    """Use morph learning progress to decorate the morphemes in the supplied text.
+    Adds css classes to the output that can be styled in the card."""
+
+    # Perf: Bail early if the user attempts to use this template filter on the already
+    # formatted data.
+    #
```

Comment:
Yes of course. 

I have been thinking about this for the last little bit now that the major hurdles are addressed......

It's worth noting that my coding "style" is pretty different from yours. 

for instance
1. I don't care for local variables if I am about to just pass them in to a function, I just inline them.
2. I use inline ifs (my primary language is Typescript and it's so much more beautiful than python) I don't think I have seen you use them elsewhere.
3. a ton of other things like these.

Since it's pretty jarring to pop open a single file in the whole repo and see these differences, I could make an attempt at making all of this more like the rest of the code in the repo.

lmk if you'd like me to do that.

Reply:
>1. I don't care for local variables if I am about to just pass them in to a function, I just inline them.

After having worked on a lot of horrible code with immense technical debts, my number one priority is code readability, unless it causes disastrous performance issues. Even the slightest improvement to readability could save an incredible amount of time in the future, so it's almost always worth the little extra effort; The fewer brain cells I have to use to understand code, the better the code is.

I think the scoring code is a good example of this:
https://github.com/mortii/anki-morphs/blob/cf3f930c9f0dddc44a0b410a7eed7a60ba1e2400/ankimorphs/recalc/card_score.py#L46-L101

If I had come across a project that had inlined all this, I would have immediately left and never looked back.

>2. I use inline ifs (my primary language is Typescript and it's so much more beautiful than python) I don't think I have seen you use them elsewhere.

Inline if-statements are usually denser, so they often fail my 'good code' eye test. The fact that the syntax order is slightly different than normal ifs also means I have to use an extra millisecond to decipher it. But it's not necessarily the end of the world, It can occasionally be cleaner. I use it in some list comprehensions I think.

> 3. a ton of other things like these.

This is an example of hard-to-read code in my opinion:
```python
return Expression(  
    text, [MorphemeMeta(morpheme, am_config) for morpheme in morphemes]  
).highlighted()
```

I can tell something Is happening, but I can't be completely confident of what it is, or how it works, by just reading this high-level statement. I would have to invest time and effort every time I come across it to understand it, which realistically means I probably won't ever look, and potential bugs and technical debt accumulates.

Even small changes like this would make a lot easier to understand, and I would be very confident that I know what is happening without having to do any further digging:
```python
morphs_and_their_statuses: list[MorphAndStatus] = [  
    MorphAndStatus(morpheme, am_config) for morpheme in morphemes  
]  
  
return TextHighlighter(original_text, morphs_and_their_statuses).highlighted()
```

Reply:
Done.